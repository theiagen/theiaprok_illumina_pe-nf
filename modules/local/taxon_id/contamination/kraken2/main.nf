process KRAKEN2_THEIACOV {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/staphb/kraken2:2.1.2-no-db"

    input:
    tuple val(meta), path(read1), path(read2)
    path kraken2_db
    val target_organism

    output:
    tuple val(meta), path("*_kraken2_report.txt"), emit: kraken_report
    tuple val(meta), path("*.classifiedreads.txt.gz"), emit: kraken2_classified_report
    tuple val(meta), path("PERCENT_HUMAN.txt"), emit: percent_human_file
    tuple val(meta), path("PERCENT_SC2.txt"), emit: percent_sc2_file
    tuple val(meta), path("PERCENT_TARGET_ORGANISM.txt"), emit: percent_target_organism_file
    tuple val(meta), val(target_organism), emit: kraken_target_organism
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def target_org = target_organism ?: ""
    def reads_input = read2 ? "${read1} ${read2}" : "${read1}"
    def mode = read2 ? "--paired" : ""
    def compression = read1.name.endsWith('.gz') ? "--gzip-compressed" : ""
    
    """
    # date and version control
    date | tee DATE
    
    # Decompress the Kraken2 database
    mkdir db
    tar -C ./db/ -xzvf ${kraken2_db}
    
    echo "Mode: ${mode}"
    echo "Compression: ${compression}"
    
    # Run Kraken2
    kraken2 ${mode} ${compression} \\
        --threads ${task.cpus} \\
        --db ./db/ \\
        ${reads_input} \\
        --report ${prefix}_kraken2_report.txt \\
        --output ${prefix}.classifiedreads.txt
    
    # Compress classified reads output
    gzip ${prefix}.classifiedreads.txt
    
    # capture human percentage
    percentage_human=\$(grep "Homo sapiens" ${prefix}_kraken2_report.txt | cut -f 1)
    if [ -z "\$percentage_human" ] ; then percentage_human="0" ; fi
    echo "\$percentage_human" > PERCENT_HUMAN.txt
    echo "DEBUG: Human percentage: \$percentage_human"
    
    # capture target org percentage
    if [ -n "${target_org}" ]; then
        echo "Target org designated: ${target_org}"
        # if target organism is sc2, report it in a special legacy column called PERCENT_SC2
        if [[ "${target_org}" == "Severe acute respiratory syndrome coronavirus 2" ]]; then
            percentage_sc2=\$(grep "Severe acute respiratory syndrome coronavirus 2" ${prefix}_kraken2_report.txt | cut -f1 )
            percent_target_organism=""
            if [ -z "\$percentage_sc2" ] ; then percentage_sc2="0" ; fi
            echo "DEBUG: SC2 percentage: \$percentage_sc2"
        else
            percentage_sc2="" 
            percent_target_organism=\$(grep "${target_org}" ${prefix}_kraken2_report.txt | cut -f1 | head -n1 )
            if [ -z "\$percent_target_organism" ] ; then percent_target_organism="0" ; fi
            echo "DEBUG: Target organism percentage: \$percent_target_organism"
        fi
    else
        percent_target_organism=""
        percentage_sc2=""
        echo "DEBUG: No target organism specified"
    fi
    
    echo "\$percentage_sc2" > PERCENT_SC2.txt
    echo "\$percent_target_organism" > PERCENT_TARGET_ORGANISM.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kraken2: \$(kraken2 --version | head -n1 | sed 's/Kraken version //')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_kraken2_report.txt
    touch ${prefix}.classifiedreads.txt.gz
    echo "0.0" > PERCENT_HUMAN.txt
    echo "0.0" > PERCENT_SC2.txt
    echo "0.0" > PERCENT_TARGET_ORGANISM.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kraken2: 2.1.2
    END_VERSIONS
    """
}