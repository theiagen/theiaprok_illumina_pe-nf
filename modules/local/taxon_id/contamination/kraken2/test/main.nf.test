nextflow_process {
    name "Test Process Kraken2"
    script "../main.nf"
    process "KRAKEN2"
    
    test("kraken2 test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = file("https://github.com/nf-core/test-datasets/raw/refs/heads/taxprofiler/data/database/kraken2/testdb-kraken2.tar.gz")  // kraken2_db
                input[2] = "" // kraken2_args
                input[3] = "classified#.fastq"  // classified_out
                input[4] = "unclassified#.fastq"  // unclassified_out
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.kraken2_report,
                    process.out.kraken2_classified_report,
                    process.out.kraken2_unclassified_reads,
                    process.out.kraken2_classified_reads,
                    process.out.kraken2_percent_human_file,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("kraken2 - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = file("https://github.com/nf-core/test-datasets/raw/refs/heads/taxprofiler/data/database/kraken2/testdb-kraken2.tar.gz") // kraken2_db
                input[2] = "" // kraken2_args
                input[3] = "classified#.fastq"  // classified_out
                input[4] = "unclassified#.fastq"  // unclassified_out
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}