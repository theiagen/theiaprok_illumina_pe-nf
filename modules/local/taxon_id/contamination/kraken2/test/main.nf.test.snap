{
    "kraken2 - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.report.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.classifiedreads.txt.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.unclassified_1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.classified_1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PERCENT_HUMAN.txt:md5,6808ca805661622ad65ae014a4b2a094"
                    ]
                ],
                "5": [
                    "versions.yml:md5,8b0b6f59777a156266d8c1123a26c1ce"
                ],
                "kraken2_classified_reads": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.classified_1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kraken2_classified_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.classifiedreads.txt.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kraken2_percent_human_file": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "PERCENT_HUMAN.txt:md5,6808ca805661622ad65ae014a4b2a094"
                    ]
                ],
                "kraken2_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.report.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kraken2_unclassified_reads": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.unclassified_1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8b0b6f59777a156266d8c1123a26c1ce"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-18T19:00:11.249733966"
    },
    "kraken2 test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.report.txt:md5,a3a38882bd402d836db72bd20ef948b3"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.classifiedreads.txt.gz:md5,d1da20259b1133a73069aebb8073410a"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample.unclassified_1.fastq.gz:md5,48ab5cdbf8567cf6023bd304fcf896a9",
                        "test_sample.unclassified_2.fastq.gz:md5,b4246225aa4a15496991ac7c6b58563f"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample.classified_1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test_sample.classified_2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "PERCENT_HUMAN.txt:md5,897316929176464ebc9ad085f31e7284"
                ]
            ],
            [
                "versions.yml:md5,8b0b6f59777a156266d8c1123a26c1ce"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-18T18:59:57.033239171"
    }
}